src/rust/src/backend/dh.rs

```diff
@@ -77,11 +77,7 @@ fn pkey_from_dh<T: openssl::pkey::HasParams>(
         if #[cfg(CRYPTOGRAPHY_IS_LIBRESSL)] {
             Ok(openssl::pkey::PKey::from_dh(dh)?)
         } else {
-            if dh.prime_q().is_some() {
-                Ok(openssl::pkey::PKey::from_dhx(dh)?)
-            } else {
-                Ok(openssl::pkey::PKey::from_dh(dh)?)
-            }
+            Ok(openssl::pkey::PKey::from_dh(dh)?)
```

Comment:
So with this change it will never create a `dhx` variant, which I think is more aggressive than we want.

I'm honestly unsure how it is that no tests fail because of that.

I think you only want to apply this change to the `generate` codepaths. My suggest would be to add an `allow_dhx` param (maybe a better name), pass false in generate, but true everywhere else.